cub/cub/device/dispatch/dispatch_transform.cuh

```diff
@@ -0,0 +1,1128 @@
+  //   https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#device-memory-accesses
+  //   > Any address of a variable residing in global memory or returned by one of the memory allocation routines from
+  //     the driver or runtime API is always aligned to at least 256 bytes.
+  // However, gevtushenko says since Linux Kernel 6, any host memory is device accessible, even stack memory
+  // ahendriksen argues that all memory pages are sufficiently aligned, even those migrated from the host
+  const T* base_ptr = round_down_ptr(ptr, alignment);
+  return aligned_base_ptr<const T>{base_ptr, static_cast<int>((ptr - base_ptr) * sizeof(T))};
+}
+
+// For performance considerations of memcpy_async:
+// https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#performance-guidance-for-memcpy-async
+
+constexpr int memcpy_async_alignment     = 16;
+constexpr int memcpy_async_size_multiple = 16;
+
+// Our own version of ::cuda::aligned_size_t, since we cannot include <cuda/barrier> on CUDA_ARCH < 700
+template <_CUDA_VSTD::size_t _Alignment>
+struct aligned_size_t
+{
+  _CUDA_VSTD::size_t value;
+
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr operator size_t() const
+  {
+    return value;
+  }
+};
+
+// TODO(bgruber): inline this as lambda in C++14
+template <typename T>
+_CCCL_DEVICE const T* copy_and_return_smem_dst(
+  cooperative_groups::thread_block& group,
+  int tile_size,
+  char* smem,
+  int& smem_offset,
+  int global_offset,
+  aligned_base_ptr<const T> aligned_ptr)
+{
+  const int count =
+    round_up_to_po2_multiple(static_cast<int>(sizeof(T)) * tile_size + aligned_ptr.offset, memcpy_async_size_multiple);
+  // because SMEM base pointer and count are always multiples of 16-byte, we only need to align the SMEM start for types
+  // with larger alignment
+  _CCCL_IF_CONSTEXPR (alignof(T) > memcpy_async_alignment)
+  {
+    smem_offset = round_up_to_po2_multiple(smem_offset, static_cast<int>(alignof(T)));
+  }
+  auto smem_dst = reinterpret_cast<T*>(smem + smem_offset);
+  assert(reinterpret_cast<uintptr_t>(smem_dst) % memcpy_async_size_multiple == 0); // to hit optimal memcpy_async
+                                                                                   // performance
+  cooperative_groups::memcpy_async(
+    group,
+    smem_dst,
+    aligned_ptr.ptr + global_offset,
+    aligned_size_t<memcpy_async_size_multiple>{static_cast<::cuda::std::size_t>(count)});
+  smem_offset += count;
+  return smem_dst + aligned_ptr.offset;
+}
+
+template <typename, typename Offset, typename F, typename RandomAccessIteartorOut, typename... InTs>
+_CCCL_DEVICE void transform_kernel_impl(
+  ::cuda::std::integral_constant<Algorithm, Algorithm::memcpy_async>,
+  Offset num_items,
+  int num_elem_per_thread,
+  F f,
+  RandomAccessIteartorOut out,
+  aligned_base_ptr<const InTs>... aligned_ptrs)
+{
+  extern __shared__ char smem[]; // this should be __attribute((aligned(memcpy_async_alignment))), but then it clashes
```

Comment:
you can even use `__builtin_assume_aligned()` to get a new address with alignment guarantees

Reply:
Note that this requires compiler support and need to be integrated not to break our support matrix

Reply:
indeed we would need  `assume_aligned` [C++20](https://en.cppreference.com/w/cpp/memory/assume_aligned)